### PR TITLE
feat(openai): assistants API

### PR DESCRIPTION
**Pull request by nirga**
- [x] I have added tests that cover my changes.
- [x] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

Screenshot 2024-03-26 at 15 03 25

### CODE REVIEW

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        return response
+
+    run = runs[id]
+    messages = sorted(response_dict["data"], key=lambda x: x["created_at"])
+
+    span = tracer.start_span(
+        "openai.assistant.run",
+        kind=SpanKind.CLIENT,
+        attributes={SpanAttributes.LLM_REQUEST_TYPE: LLMRequestTypeValues.CHAT.value},
+        start_time=run.get("start_time"),
```
Comment (galkleinman) [inline packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py:91]: potential mess

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+
+@_with_tracer_wrapper
+def runs_create_and_stream_wrapper(tracer, wrapped, instance, args, kwargs):
+    if context_api.get_value(_SUPPRESS_INSTRUMENTATION_KEY):
+        return wrapped(*args, **kwargs)
+
+    from openai import AssistantEventHandler
+
+    class EventHandleWrapper(AssistantEventHandler):
```
Comment (galkleinman) [inline packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py:134]: [STYLE] move to a dedicated file?

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+from opentelemetry.instrumentation.utils import _SUPPRESS_INSTRUMENTATION_KEY
+
+from opentelemetry.semconv.ai import SpanAttributes, LLMRequestTypeValues
+
+from opentelemetry.instrumentation.openai.utils import _with_tracer_wrapper
+
+logger = logging.getLogger(__name__)
+
+assistants = {}
+runs = {}
```
Comment (galkleinman) [inline packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py:19]: wonder how do these ones behave in an async code

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py (hunk 1 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py:91]: 🤷‍♂️

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py (hunk 3 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/assistant_wrappers.py:18]: I think yes
https://stackoverflow.com/questions/65041691/is-python-dictionary-async-safe

**Review by galkleinman — APPROVED** [review]

**Pull request merged**